File: mobile/library/swift/grpc/GRPCStreamPrototype.swift
import Foundation

/// A type representing a gRPC stream that has not yet been started.
///
/// Constructed via `GRPCClient`, and used to assign response callbacks
/// prior to starting a `GRPCStream` by calling `start()`.
@objcMembers
public final class GRPCStreamPrototype: NSObject {
  private let underlyingStream: StreamPrototype

  /// Initialize a new instance of the inactive gRPC stream.
  ///
  /// - parameter underlyingStream: The underlying stream to use.
  required init(underlyingStream: StreamPrototype) {
    self.underlyingStream = underlyingStream
    super.init()
  }

  // MARK: - Public

  /// Start a new gRPC stream.
  ///
  /// - parameter queue: Queue on which to receive callback events.
  ///
  /// - returns: The new gRPC stream.
  public func start(queue: DispatchQueue = .main) -> GRPCStream {
    let stream = self.underlyingStream.start(queue: queue)
    return GRPCStream(underlyingStream: stream)
  }

  /// Specify a callback for when response headers are received by the stream.
  ///
  /// - parameter closure: Closure which will receive the headers
  ///                      and flag indicating if the stream is headers-only.
  ///
  /// - returns: This stream, for chaining syntax.
  @discardableResult
  public func setOnResponseHeaders(
    closure: @escaping (_ headers: ResponseHeaders, _ endStream: Bool,
                        _ streamIntel: StreamIntel) -> Void
  ) -> GRPCStreamPrototype {
    self.underlyingStream.setOnResponseHeaders(closure: closure)
    return self
  }

  /// Specify a callback for when a new message has been received by the stream.
  ///
  /// - parameter closure: Closure which will receive messages on the stream.
  ///
  /// - returns: This handler, which may be used for chaining syntax.
  @discardableResult
  public func setOnResponseMessage(
    _ closure: @escaping (_ message: Data, _ streamIntel: StreamIntel) -> Void
  ) -> GRPCStreamPrototype {
    var buffer = Data()
    var state = GRPCMessageProcessor.State.expectingCompressionFlag
    self.underlyingStream.setOnResponseData { chunk, _, streamIntel in
      // Appending might result in extra copying that can be optimized in the future.
      buffer.append(chunk)
      // gRPC always sends trailers, so the stream will not complete here.
      GRPCMessageProcessor.processBuffer(&buffer, state: &state, streamIntel: streamIntel,
                                         onMessage: closure)
    }

    return self
  }

  /// Specify a callback for when trailers are received by the stream.
  /// If the closure is called, the stream is complete.
  ///
  /// - parameter closure: Closure which will receive the trailers.
  ///
  /// - returns: This handler, which may be used for chaining syntax.
  @discardableResult
  public func setOnResponseTrailers(_ closure:
    @escaping (_ trailers: ResponseTrailers, _ streamIntel: StreamIntel) -> Void
  ) -> GRPCStreamPrototype {
    self.underlyingStream.setOnResponseTrailers(closure: closure)
    return self
  }

  /// Specify a callback for when an internal Envoy exception occurs with the stream.
  /// If the closure is called, the stream is complete.
  ///
  /// - parameter closure: Closure which will be called when an error occurs.
  ///
  /// - returns: This handler, which may be used for chaining syntax.
  @discardableResult
  public func setOnError(
    _ closure: @escaping (_ error: EnvoyError, _ streamIntel: StreamIntel) -> Void
  ) -> GRPCStreamPrototype {
    self.underlyingStream.setOnError(closure: closure)
    return self
  }

  /// Specify a callback for when the stream is canceled.
  /// If the closure is called, the stream is complete.
  ///
  /// - parameter closure: Closure which will be called when the stream is canceled.
  ///
  /// - returns: This stream, for chaining syntax.
  @discardableResult
  public func setOnCancel(
    closure: @escaping (_ streamInte: StreamIntel) -> Void
  ) -> GRPCStreamPrototype {
    self.underlyingStream.setOnCancel(closure: closure)
    return self
  }
}

private enum GRPCMessageProcessor {
  /// Represents the state of a response stream's body data.
  enum State {
    /// Awaiting a gRPC compression flag.
    case expectingCompressionFlag
    /// Awaiting the length specification of the next message.
    case expectingMessageLength
    /// Awaiting a message with the specified length.
    case expectingMessage(messageLength: UInt32)
  }

  /// Recursively processes a buffer of data, buffering it into messages based on state.
  /// When a message has been fully buffered, `onMessage` will be called with the message.
  ///
  /// - parameter buffer:    The buffer of data from which to determine state and messages.
  /// - parameter state:     The current state of the buffering.
  /// - parameter onMessage: Closure to call when a new message is available.
  static func processBuffer(_ buffer: inout Data, state: inout State, streamIntel: StreamIntel,
                            onMessage: (_ message: Data, _ streamIntel: StreamIntel) -> Void)
  {
    switch state {
    case .expectingCompressionFlag:
      guard let compressionFlag: UInt8 = buffer.integer(atIndex: 0) else {
        return
      }

      guard compressionFlag == 0 else {
        // TODO: Support gRPC compression https://github.com/envoyproxy/envoy-mobile/issues/501
        buffer.removeAll()
        state = .expectingCompressionFlag
        return
      }

      state = .expectingMessageLength

    case .expectingMessageLength:
      guard let messageLength: UInt32 = buffer.integer(atIndex: 1) else {
        return
      }

      state = .expectingMessage(messageLength: CFSwapInt32BigToHost(messageLength))

    case .expectingMessage(let messageLength):
      let prefixedLength = kGRPCPrefixLength + Int(messageLength)
      if buffer.count < prefixedLength {
        return
      }

      if messageLength > 0 {
        onMessage(buffer.subdata(in: kGRPCPrefixLength..<prefixedLength), streamIntel)
      } else {
        onMessage(Data(), streamIntel)
      }

      buffer.removeSubrange(0..<prefixedLength)
      state = .expectingCompressionFlag
    }

    self.processBuffer(&buffer, state: &state, streamIntel: streamIntel, onMessage: onMessage)
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/grpc/GRPCStreamPrototype.kt
package io.envoyproxy.envoymobile

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * A type representing a gRPC stream that has not yet been started.
 *
 * Constructed via `GRPCClient`, and used to assign response callbacks prior to starting a
 * `GRPCStream` by calling `start()`.
 */
class GRPCStreamPrototype(private val underlyingStream: StreamPrototype) {
  /**
   * Start a new gRPC stream.
   *
   * @param executor Executor on which to receive callback events.
   * @return The new gRPC stream.
   */
  fun start(executor: Executor = Executors.newSingleThreadExecutor()): GRPCStream {
    val stream = underlyingStream.start(executor)
    return GRPCStream(stream)
  }

  /**
   * Specify a callback for when response headers are received by the stream.
   *
   * @param closure Closure which will receive the headers and flag indicating if the stream is
   *   headers-only.
   * @return This stream, for chaining syntax.
   */
  fun setOnResponseHeaders(
    closure: (headers: ResponseHeaders, endStream: Boolean, streamIntel: StreamIntel) -> Unit
  ): GRPCStreamPrototype {
    underlyingStream.setOnResponseHeaders(closure)
    return this
  }

  /**
   * Specify a callback for when a new message has been received by the stream. If `endStream` is
   * `true`, the stream is complete.
   *
   * @param closure Closure which will receive messages on the stream.
   * @return This stream, for chaining syntax.
   */
  fun setOnResponseMessage(
    closure: (data: ByteBuffer, streamIntel: StreamIntel) -> Unit
  ): GRPCStreamPrototype {
    val byteBufferedOutputStream = ByteArrayOutputStream()
    val processor = GRPCMessageProcessor()
    var processState: GRPCMessageProcessor.ProcessState =
      GRPCMessageProcessor.ProcessState.CompressionFlag
    underlyingStream.setOnResponseData { byteBuffer, _, streamIntel ->
      val byteBufferArray =
        if (byteBuffer.hasArray()) {
          byteBuffer.array()
        } else {
          val array = ByteArray(byteBuffer.remaining())
          byteBuffer.get(array)
          array
        }
      byteBufferedOutputStream.write(byteBufferArray)

      processState =
        processor.processData(byteBufferedOutputStream, processState, streamIntel, closure)
    }

    return this
  }

  /**
   * Specify a callback for when trailers are received by the stream. If the closure is called, the
   * stream is complete.
   *
   * @param closure Closure which will receive the trailers.
   * @return This stream, for chaining syntax.
   */
  fun setOnResponseTrailers(
    closure: (trailers: ResponseTrailers, streamIntel: StreamIntel) -> Unit
  ): GRPCStreamPrototype {
    underlyingStream.setOnResponseTrailers(closure)
    return this
  }

  /**
   * Specify a callback for when an internal Envoy exception occurs with the stream. If the closure
   * is called, the stream is complete.
   *
   * @param closure Closure which will be called when an error occurs.
   * @return This stream, for chaining syntax.
   */
  fun setOnError(
    closure: (error: EnvoyError, finalStreamIntel: FinalStreamIntel) -> Unit
  ): GRPCStreamPrototype {
    underlyingStream.setOnError(closure)
    return this
  }

  /**
   * Specify a callback for when the stream is canceled. If the closure is called, the stream is
   * complete.
   *
   * @param closure Closure which will be called when the stream is canceled.
   * @return This stream, for chaining syntax.
   */
  fun setOnCancel(closure: (finalStreamIntel: FinalStreamIntel) -> Unit): GRPCStreamPrototype {
    underlyingStream.setOnCancel(closure)
    return this
  }
}

private class GRPCMessageProcessor {
  /** Represents the process state of the response stream's body data. */
  sealed class ProcessState {
    // Awaiting a gRPC compression flag.
    object CompressionFlag : ProcessState()

    // Awaiting the length specification of the next message.
    object MessageLength : ProcessState()

    // Awaiting a message with the specified length.
    class Message(val messageLength: Int) : ProcessState()
  }

  /**
   * Recursively processes a buffer of data, buffering it into messages based on state. When a
   * message has been fully buffered, `onMessage` will be called with the message.
   *
   * @param bufferedStream The buffer of data from which to determine state and messages.
   * @param processState The current process state of the buffering.
   * @param onMessage Closure to call when a new message is available.
   * @return The state after processing the passed data.
   */
  fun processData(
    bufferedStream: ByteArrayOutputStream,
    processState: GRPCMessageProcessor.ProcessState,
    streamIntel: StreamIntel,
    onMessage: (byteBuffer: ByteBuffer, streamIntel: StreamIntel) -> Unit
  ): GRPCMessageProcessor.ProcessState {
    var nextState = processState

    when (processState) {
      is ProcessState.CompressionFlag -> {
        val byteArray = bufferedStream.toByteArray()
        if (byteArray.isEmpty()) {
          // We don't have enough information to extract the compression flag, so we'll just return
          return ProcessState.CompressionFlag
        }

        val compressionFlag = byteArray[0]
        // TODO: Support gRPC compression https://github.com/envoyproxy/envoy-mobile/issues/501.
        if (compressionFlag.compareTo(0) != 0) {
          bufferedStream.reset()
        }

        nextState = ProcessState.MessageLength
      }
      is ProcessState.MessageLength -> {
        if (bufferedStream.size() < GRPC_PREFIX_LENGTH) {
          // We don't have enough information to extract the message length, so we'll just return
          return ProcessState.MessageLength
        }

        val byteArray = bufferedStream.toByteArray()
        val buffer = ByteBuffer.wrap(byteArray.sliceArray(1..4))
        buffer.order(ByteOrder.BIG_ENDIAN)
        val messageLength = buffer.int
        nextState = ProcessState.Message(messageLength)
      }
      is ProcessState.Message -> {
        if (bufferedStream.size() < processState.messageLength + GRPC_PREFIX_LENGTH) {
          // We don't have enough bytes to construct the message, so we'll just return
          return ProcessState.Message(processState.messageLength)
        }

        val byteArray = bufferedStream.toByteArray()
        onMessage(
          ByteBuffer.wrap(
            byteArray.sliceArray(
              GRPC_PREFIX_LENGTH until GRPC_PREFIX_LENGTH + processState.messageLength
            )
          ),
          streamIntel
        )
        bufferedStream.reset()
        bufferedStream.write(
          byteArray.sliceArray(GRPC_PREFIX_LENGTH + processState.messageLength until byteArray.size)
        )

        val remainingLength = GRPC_PREFIX_LENGTH + processState.messageLength until byteArray.size
        if (byteArray.sliceArray(remainingLength).isEmpty()) {
          return ProcessState.CompressionFlag
        } else {
          nextState = ProcessState.CompressionFlag
        }
      }
    }

    return processData(bufferedStream, nextState, streamIntel, onMessage)
  }
}
